contexts/DonationContext/src/UseCases/HandlePayPalPaymentNotification/HandlePayPalPaymentNotificationUseCase.php

```diff
@@ -61,82 +61,59 @@ public function handleNotification( PayPalNotificationRequest $request ): bool {
 		$donation = $this->newDonationFromRequest( $request );
 
 		try {
 			$this->repository->storeDonation( $donation );
 		} catch ( StoreDonationException $ex ) {
-			return false;
+			return $this->createErrorResponse( $ex, $request );
 		}
 
 		$this->sendConfirmationEmailFor( $donation );
 		$this->donationEventLogger->log( $donation->getId(), 'paypal_handler: booked' );
 
-		return true;
+		return PaypalNotificationResponse::newSuccessResponse();
 	}
 
-	private function handleRequestForDonation( PayPalNotificationRequest $request, Donation $donation ): bool {
+	private function handleRequestForDonation( PayPalNotificationRequest $request, Donation $donation ): PaypalNotificationResponse {
 		if ( !( $donation->getPayment()->getPaymentMethod() instanceof PayPalPayment ) ) {
-			return false;
+			return $this->createUnhandledResponse( 'Trying to handle IPN for non-Paypal donation', $request );
 		}
 
 		if ( !$this->authorizationService->systemCanModifyDonation( $request->getDonationId() ) ) {
-			return false;
+			return $this->createUnhandledResponse( 'Wrong access code for donation', $request );
 		}
 		if ( $this->donationWasBookedWithDifferentTransactionId( $donation, $request ) ) {
-			$childDonation = $this->createChildDonation( $donation, $request );
-			return $childDonation !== null;
+			return $this->createChildDonation( $donation, $request );
 		}
 
 		$donation->addPayPalData( $this->newPayPalDataFromRequest( $request ) );
 
 		try {
 			$donation->confirmBooked();
 		} catch ( \RuntimeException $ex ) {
-			return false;
+			return $this->createErrorResponse( $ex, $request );
 		}
 
 		try {
 			$this->repository->storeDonation( $donation );
 		}
 		catch ( StoreDonationException $ex ) {
-			return false;
+			return $this->createErrorResponse( $ex, $request );
 		}
 
 		$this->sendConfirmationEmailFor( $donation );
 		$this->donationEventLogger->log( $donation->getId(), 'paypal_handler: booked' );
 
-		return true;
-	}
-
-	private function logUnhandledStatus( PayPalNotificationRequest $request ) {
-		$logContext = [
-			'payment_status' => $request->getPaymentStatus(),
-			'txn_id' => $request->getTransactionId()
-		];
-		$this->logger->info( 'Unhandled PayPal notification: ' . $request->getPaymentStatus(), $logContext );
+		return PaypalNotificationResponse::newSuccessResponse();
 	}
 
-	private function logUnhandledNonPayment( PayPalNotificationRequest $request ) {
-		$logContext = [
-			'transaction_type' => $request->getTransactionType(),
-			'txn_id' => $request->getTransactionId()
-		];
-		$this->logger->info( 'Unhandled PayPal subscription notification: ' . $request->getTransactionType(), $logContext );
+	private function createUnhandledResponse( string $reason, PayPalNotificationRequest $request ): PaypalNotificationResponse {
+		return PaypalNotificationResponse::newUnhandledResponse( [
+			'message' => $reason,
+			'request' => $request->toArray()
```

Comment:
Agree. The intention behind that was to check what PayPal sent in its request, but now that I've looked at it again I think it would be much better to get rid of `toArray` and the `request` field in the response and add the parameters directly from the HTTP request when logging in `PayPalNotificationHandler`.